# Panagram: Interactive, alignment-free pan-genome browser  

#### Katie Jenike, Sam Kovaka, Shujun Ou, Stephen Hwang, Srividya Ramakrishnan, Ben Langmead, Zach Lippman, Michael Schatz


[An alignment-free pan-genome viewer](https://www.dropbox.com/s/g7snjgr8bs6c2uj/2023.01.17.Panagram.pdf)

Please note: installation instructions and pre-processing scripts are a work in progress. 

# Installation

```
> git clone --recursive https://github.com/kjenike/panagram.git
> cd panagram
> pip install .
```

Panagram relies on [KMC](https://github.com/refresh-bio/KMC) to build its kmer index. This should be installed automatically, however it is possible that the KMC installation will fail but panagram will successfully install. In this case `panagram view` can be run, but `panagram index` will return an error. You may be able to debug the KMC installation by running `make -C KMC py_kmc_api` and attempting to fix any errors, then re-run `pip install -v .` after the errors are fixed.

# Running
Panagram runs in two steps, the pre-processing step (index command) and the viewing (view command). 

# Preprocessing
Usage:
Anchor KMC bitvectors to reference FASTA files to create pan-kmer bitmap
```
usage: panagram index [-h] <config>.toml
```
See example config.toml file for more details on the layout. Must include paths to all of the fasta files and optionally any annotations in gff format. 

# View

Usage:
```
usage: panagram view [-h] <path_to_panagram_index_directory>
  Display panagram viewer
```

Runs a local Dash server. Browser can be viewed at http://127.0.0.1:8050/ by default.

# Bitdump

Usage:
```
usage: panagram bitdump [-h] [-v bool] index_dir coords step
  Query pan-kmer bitmap generated by "panagram index"/

  index_dir             Panagram index directory
  coords                Coordinates to query in chr:start-end format
  step                  Spacing between output kmers (optimized for multiples
                        of 100) (default: 1)
  -v bool, --verbose bool
                        Output the full bitmap (default: False)
```

# Example run

First download the example_data.zip bacterial data from: 
[http://data.schatz-lab.org/panagram/](http://data.schatz-lab.org/panagram/)

[Direct link](https://bx.bio.jhu.edu/data/panagram/example_data.zip)

Unzip the archive and you will find 5 bacterial genomes plus their annotations
```
unzip example_data.zip
```

To run, first index the genomes:

```
cd example_data
panagram index conf.toml
```

Then you can panagram to visualize (from the example_data directory):
```
panagram view . 
```

From there, you can view the results in your webbrowser at [http://127.0.0.1:8050/](http://127.0.0.1:8050)


# Hosting and Proxies

Panagram uses [Dash](https://dash.plotly.com/introduction) to serve the plotly visualizations. 
Normally Dash will serve this on a dedicated webserver on port 8050 that is only viewable on 
your localhost, but you can reverse proxy to a different port and path using a web engine 
such as [nginx](https://www.nginx.com/)

For nginx, first reconfigure your nginx configuration file to add (note to be very careful 
with the use of the slash ('/') character):

```
    location /panagram {
      proxy_pass http://127.0.0.1:8050;
    }
```

The retart nginx with 

```
systemctl stop nginx
systemctl start nginx
```

Currently (Feb 5, 2023) you will also need to edit the panagram source code to support proxying.
The easiest way is to edit panagram/view.py in the panagram source code directory before
installation with `pip install`. You can also edit the installed version (e.g. ~/.local/lib/python3.8/site-packages/panagram/view.py).

First change this line (line 1597)
```
    app = dash.Dash(external_stylesheets=["https://www.w3schools.com/w3css/4/w3.css"])
```

To read:
```
app = dash.Dash(external_stylesheets=["https://www.w3schools.com/w3css/4/w3.css"],
                    url_base_pathname='/panagram/') 
```

And then edit this line (at the end of the file)
```
    app.run_server(debug=True)
```

To read:

```
    app.run_server(host='127.0.0.1', port=8050)
```

Finally you will need to run panagram using `panagram view <dir>`. You will probably want to run this in a loop
in case it needs to be restarted, such as:

```
until panagram view .; do echo "restarting"; sleep 1; done
```

We will optimize this process in future releases.

## More information coming soon!
