import sys
import argparse
import os
import cProfile

import toml
from simple_parsing import ArgumentParser, field
import dataclasses
from typing import Any, List, Tuple, Type, Union

@dataclasses.dataclass
class View:
    """Display panagram viewer"""
    config: str = field(positional=True)

    def run(self):
        from .view import view
        view(self.config)

@dataclasses.dataclass
class KMC:
    """Parameters for KMC kmer counting"""
    memory: int = field(default=8, dest="main.kmc.memory")
    processes: int = field(default=4, dest="main.kmc.processes")
    threads: int = field(default=1, dest="main.kmc.threads")

@dataclasses.dataclass
class Index:
    """Anchor KMC bitvectors to reference FASTA files to create pan-kmer bitmap"""

    conf: str = field(positional=True, metavar="confg_file")
    #genomes: str = field(positional=True)
    #"""TSV file with each genome ID in the first column and the path to a gzipped fasta in the second column"""

    kmc: KMC = KMC()

    k: int = field(alias=["-k"], default=21)
    """K-mer length (must be same as KMC database)"""

    processes: int = field(alias=["-p"], default=1)
    """Number of processes"""

    threads: int = field(default=1,help=argparse.SUPPRESS)
    memory: int = field(default=1,help=argparse.SUPPRESS)

    anchor_only: bool = False
    """Assume KMC databases already exist"""

    bitmap_resolutions: list = field(default_factory=lambda:[1, 100])

    def _load_dict(self, d, tgt):
        for key,val in d.items():
            if isinstance(val, dict) and dataclasses.is_dataclass(getattr(tgt, key, None)):
                self._load_dict(val, getattr(tgt, key))
            else:
                setattr(tgt, key, val)

    def run(self):
        from .index import index
        args = dataclasses.asdict(self)
        del args["conf"]
        self._load_dict(toml.load(self.conf), self)

        index(conf=self)#self.genomes, self.out_dir, self.k)

@dataclasses.dataclass
class Bitdump:
    """Query pan-kmer bitmap generated by "panagram index" """

    index_dir: str = field(positional=True)
    """Panagram index directory"""

    coords: str = field(positional=True)
    """Coordinates to query in chr:start-end format"""

    step: int = field(positional=True, nargs="?", default=1)
    """Spacing between output kmers (optimized for multiples of 100)"""

    verbose: bool = field(alias=["-v"], default=False)
    """Output the full bitmap"""

    def run(self):
        from .index import Index #KmerBitmap
        bitmap = Index(self.index_dir)
        genome, chrom, start, end = parse_coords(self.coords)
        bits = bitmap.query_bitmap(genome, chrom, start, end, self.step)

        if self.verbose:
            print(" ".join(bitmap.genome_names))
            for i in range(len(bits)):
                print(" ".join(bits[i].astype(str)))
        else:
            print(bits)

        #g = bitmap.genomes

        bitmap.close()

@dataclasses.dataclass
class Main:
    """Alignment-free pan-genome viewer

Subcommands:
    index    Anchor KMC bitvectors to reference FASTA files
    view     Display panagram viewer in a browser window
    bitdump  Query pan-kmer bitmap via the commandline"""

    cmd: Union[View, Index, Bitdump]
    cprof: str = field(default=None, help=argparse.SUPPRESS)

    def run(self):
        return self.cmd.run()

def parse_coords(coords):
    genome, chrom, coords = coords.split(":")
    start, end = map(int, coords.split("-"))
    return genome, chrom, start, end

def comma_split(s):
    return s.split(",")

def main():
    parser = ArgumentParser()
    parser.add_arguments(Main, dest="main")
    args = parser.parse_args()
    if args.main.cprof is None:
        args.main.run()
    else:
        cProfile.runctx("args.run()",
             {},{"args" : args},
             args.main.cprof)

